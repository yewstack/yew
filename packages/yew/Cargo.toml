[package]
name = "yew"
version = "0.17.4"
edition = "2018"
authors = [
  "Denis Kolodin <deniskolodin@gmail.com>",
  "Justin Starry <justin@yew.rs>",
]
repository = "https://github.com/yewstack/yew"
homepage = "https://github.com/yewstack/yew"
documentation = "https://docs.rs/yew/"
license = "MIT/Apache-2.0"
readme = "../README.md"
keywords = ["web", "webasm", "javascript"]
categories = ["gui", "wasm", "web-programming"]
description = "A framework for making client-side single-page apps"

[dependencies]
anyhow = "1"
anymap = "0.12"
console_error_panic_hook = "0.1"
gloo = "0.2.1"
http = "0.2"
indexmap = { version = "1.5", features = ["std"] }
js-sys = "0.3"
log = "0.4"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
slab = "0.4"
thiserror = "1"
wasm-bindgen = "0.2.60"
wasm-bindgen-futures = "0.4"
yew-macro = { version = "0.17.0", path = "../yew-macro" }

# optional encodings
bincode = { version = "1", optional = true }
rmp-serde = { version = "0.15.0", optional = true }
serde_cbor = { version = "0.11.1", optional = true }
serde_yaml = { version = "0.8.4", optional = true }
toml = { version = "0.5", optional = true }

[dependencies.web-sys]
version = "0.3"
features = [
  "AbortController",
  "AbortSignal",
  "AnimationEvent",
  "BinaryType",
  "Blob",
  "BlobPropertyBag",
  "console",
  "DedicatedWorkerGlobalScope",
  "Document",
  "DomTokenList",
  "DragEvent",
  "Element",
  "ErrorEvent",
  "Event",
  "EventTarget",
  "File",
  "FileList",
  "FileReader",
  "FocusEvent",
  "Headers",
  "HtmlElement",
  "HtmlButtonElement",
  "HtmlInputElement",
  "HtmlSelectElement",
  "HtmlTextAreaElement",
  "InputEvent",
  "KeyboardEvent",
  "Location",
  "MessageEvent",
  "MouseEvent",
  "Node",
  "ObserverCallback",
  "PointerEvent",
  "ProgressEvent",
  "ReferrerPolicy",
  "Request",
  "RequestCache",
  "RequestCredentials",
  "RequestInit",
  "RequestMode",
  "RequestRedirect",
  "Response",
  "Storage",
  "Text",
  "TouchEvent",
  "TransitionEvent",
  "UiEvent",
  "Url",
  "WebSocket",
  "WheelEvent",
  "Window",
  "Worker",
  "WorkerGlobalScope",
  "WorkerOptions",
]

[dev-dependencies]
base64 = "0.13.0"
bincode = "1"
easybench-wasm = "0.2.1"
rmp-serde = "0.15.0"
rustversion = "1.0"
serde_derive = "1"
ssri = "6.0.0"
trybuild = "1.0"
wasm-bindgen-test = "0.3.4"

[features]
default = ["agent"]
doc_test = []
wasm_test = []
wasm_bench = []
agent = ["bincode"]
yaml = ["serde_yaml"]
msgpack = ["rmp-serde"]
cbor = ["serde_cbor"]

[package.metadata.docs.rs]
features = ["yaml", "cbor", "toml", "msgpack", "doc_test"]
