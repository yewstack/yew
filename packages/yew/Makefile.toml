[env]
CARGO_MAKE_CLIPPY_ARGS = "--features cbor,msgpack,toml,yaml -- --deny=warnings"

[tasks.test]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "@@split(YEW_TEST_FLAGS, )",
    "--",
    "--features",
    "${YEW_TEST_FEATURES}",
]

[tasks.doc-test]
clear = true
run_task = { name = ["doc-test-normal"], fork = true }

[tasks.doc-test-normal]
command = "cargo"
args = [
    "test",
    "--doc",
    "--features",
    "doc_test,wasm_test,yaml,msgpack,cbor,toml",
]

[tasks.bench]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "--release",
    "--firefox",
    "--headless",
    "--",
    "--features",
    "wasm_bench",
    "bench",
]
