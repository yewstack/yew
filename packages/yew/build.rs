use std::env;

pub fn main() {
    let using_web_sys = cfg!(feature = "web_sys");
    let using_std_web = cfg!(feature = "std_web");
    if using_web_sys && using_std_web {
        panic!("Yew does not allow the `web_sys` and `std_web` cargo features to be used simultaneously");
    } else if !using_web_sys && !using_std_web {
        panic!("Yew requires selecting either the `web_sys` or `std_web` cargo feature");
    }

    let using_cargo_web = env::var("COMPILING_UNDER_CARGO_WEB").is_ok();
    if using_cargo_web && using_web_sys {
        panic!("cargo-web is not compatible with web-sys");
    }
}
