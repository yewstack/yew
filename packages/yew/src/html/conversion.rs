use super::{Component, NodeRef, Scope};
use std::{borrow::Cow, rc::Rc};

/// Marker trait for types that the [`html!`] macro may clone implicitly.
pub trait ImplicitClone: Clone {}

// this is only implemented because there's no way to avoid cloning this value
impl ImplicitClone for Cow<'static, str> {}

impl<T: ImplicitClone> ImplicitClone for Option<T> {}
impl<T> ImplicitClone for Rc<T> {}

impl ImplicitClone for NodeRef {}
impl<Comp: Component> ImplicitClone for Scope<Comp> {}
// TODO there are still a few missing like AgentScope

/// A trait similar to `Into<T>` which allows conversion to a value of a `Properties` struct.
pub trait IntoPropValue<T> {
    /// Convert `self` to a value of a `Properties` struct.
    fn into_prop_value(self) -> T;
}

impl<T> IntoPropValue<T> for T {
    #[inline]
    fn into_prop_value(self) -> T {
        self
    }
}
impl<T> IntoPropValue<T> for &T
where
    T: ImplicitClone,
{
    #[inline]
    fn into_prop_value(self) -> T {
        self.clone()
    }
}

impl<T> IntoPropValue<Option<T>> for T {
    #[inline]
    fn into_prop_value(self) -> Option<T> {
        Some(self)
    }
}
impl<T> IntoPropValue<Option<T>> for &T
where
    T: ImplicitClone,
{
    #[inline]
    fn into_prop_value(self) -> Option<T> {
        Some(self.clone())
    }
}

macro_rules! impl_into_prop {
    (|$value:ident: $from_ty:ty| -> $to_ty:ty { $conversion:expr }) => {
        // implement V -> T
        impl IntoPropValue<$to_ty> for $from_ty {
            #[inline]
            fn into_prop_value(self) -> $to_ty {
                let $value = self;
                $conversion
            }
        }
        // implement V -> Option<T>
        impl IntoPropValue<Option<$to_ty>> for $from_ty {
            #[inline]
            fn into_prop_value(self) -> Option<$to_ty> {
                let $value = self;
                Some({ $conversion })
            }
        }
        // implement Option<V> -> Option<T>
        impl IntoPropValue<Option<$to_ty>> for Option<$from_ty> {
            #[inline]
            fn into_prop_value(self) -> Option<$to_ty> {
                self.map(IntoPropValue::into_prop_value)
            }
        }
    };
}

// implemented with literals in mind
impl_into_prop!(|value: &'static str| -> String { value.to_owned() });

impl_into_prop!(|value: &'static str| -> Cow<'static, str> { Cow::Borrowed(value) });
impl_into_prop!(|value: String| -> Cow<'static, str> { Cow::Owned(value) });

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_str() {
        let _: String = "foo".into_prop_value();
        let _: Option<String> = "foo".into_prop_value();
        let _: Cow<'static, str> = "foo".into_prop_value();
        let _: Option<Cow<'static, str>> = "foo".into_prop_value();
    }
}
