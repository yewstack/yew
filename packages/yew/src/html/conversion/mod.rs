mod into_prop_value;
pub use into_prop_value::*;
