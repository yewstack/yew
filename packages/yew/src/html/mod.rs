//! The main html module which defines components, listeners, and class helpers.

mod classes;
mod component;
mod listener;

pub use classes::*;
pub use component::*;
pub use listener::*;

use crate::virtual_dom::VNode;
use cfg_if::cfg_if;
use cfg_match::cfg_match;
use std::cell::RefCell;
use std::rc::Rc;
cfg_if! {
    if #[cfg(feature = "std_web")] {
        use stdweb::unstable::TryFrom;
        use stdweb::web::Node;
    } else if #[cfg(feature = "web_sys")] {
        use wasm_bindgen::JsValue;
        use web_sys::Node;
    }
}

/// A type which expected as a result of `view` function implementation.
pub type Html = VNode;

/// Wrapped Node reference for later use in Component lifecycle methods.
///
/// # Example
/// Focus an `<input>` element on mount.
/// ```
/// #[cfg(feature = "std_web")]
/// use stdweb::web::{html_element::InputElement, IHtmlElement};
/// #[cfg(feature = "web_sys")]
/// use web_sys::HtmlInputElement as InputElement;
///# use yew::prelude::*;
///
/// pub struct Input {
///     node_ref: NodeRef,
/// }
///
/// impl Component for Input {
///     type Message = ();
///     type Properties = ();
///
///     fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self {
///         Input {
///             node_ref: NodeRef::default(),
///         }
///     }
///
///     fn rendered(&mut self, first_render: bool) {
///         if first_render {
///             if let Some(input) = self.node_ref.cast::<InputElement>() {
///                 input.focus();
///             }
///         }
///     }
///
///     fn change(&mut self, _: Self::Properties) -> ShouldRender {
///         false
///     }
///
///     fn update(&mut self, _: Self::Message) -> ShouldRender {
///         false
///     }
///
///     fn view(&self) -> Html {
///         html! {
///             <input ref=self.node_ref.clone() type="text" />
///         }
///     }
/// }
/// ```
/// ## Relevant examples
/// - [Node Refs](https://github.com/yewstack/yew/tree/master/examples/node_refs)
#[derive(Debug, Default, Clone)]
pub struct NodeRef(Rc<RefCell<NodeRefInner>>);

impl PartialEq for NodeRef {
    fn eq(&self, other: &Self) -> bool {
        self.0.as_ptr() == other.0.as_ptr() || Some(self) == other.0.borrow().link.as_ref()
    }
}

#[derive(PartialEq, Debug, Default, Clone)]
struct NodeRefInner {
    node: Option<Node>,
    link: Option<NodeRef>,
}

impl NodeRef {
    /// Get the wrapped Node reference if it exists
    pub fn get(&self) -> Option<Node> {
        let inner = self.0.borrow();
        inner.node.clone().or_else(|| inner.link.as_ref()?.get())
    }

    /// Try converting the node reference into another form
    pub fn cast<
        #[cfg(feature = "std_web")] INTO: TryFrom<Node>,
        #[cfg(feature = "web_sys")] INTO: AsRef<Node> + From<JsValue>,
    >(
        &self,
    ) -> Option<INTO> {
        let node = self.get();
        cfg_match! {
            feature = "std_web" => node.and_then(|node| INTO::try_from(node).ok()),
            feature = "web_sys" => node.map(Into::into).map(INTO::from),
        }
    }

    /// Wrap an existing `Node` in a `NodeRef`
    pub(crate) fn new(node: Node) -> Self {
        let node_ref = NodeRef::default();
        node_ref.set(Some(node));
        node_ref
    }

    /// Place a Node in a reference for later use
    pub(crate) fn set(&self, node: Option<Node>) {
        let mut this = self.0.borrow_mut();
        this.node = node;
        this.link = None;
    }

    /// Link a downstream `NodeRef`
    pub(crate) fn link(&self, node_ref: Self) {
        // Avoid circular references
        if self == &node_ref {
            return;
        }

        let mut this = self.0.borrow_mut();
        this.node = None;
        this.link = Some(node_ref);
    }

    /// Reuse an existing `NodeRef`
    pub(crate) fn reuse(&self, node_ref: Self) {
        // Avoid circular references
        if self == &node_ref {
            return;
        }

        let mut this = self.0.borrow_mut();
        let existing = node_ref.0.borrow();
        this.node = existing.node.clone();
        this.link = existing.link.clone();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::document;

    #[cfg(feature = "wasm_test")]
    use wasm_bindgen_test::{wasm_bindgen_test as test, wasm_bindgen_test_configure};

    #[cfg(feature = "wasm_test")]
    wasm_bindgen_test_configure!(run_in_browser);

    #[test]
    fn self_linking_node_ref() {
        let node: Node = document().create_text_node("test node").into();
        let node_ref = NodeRef::new(node.clone());
        let node_ref_2 = NodeRef::new(node.clone());

        // Link to self
        node_ref.link(node_ref.clone());
        assert_eq!(node, node_ref.get().unwrap());

        // Create cycle of two node refs
        node_ref.link(node_ref_2.clone());
        node_ref_2.link(node_ref);
        assert_eq!(node, node_ref_2.get().unwrap());
    }
}
