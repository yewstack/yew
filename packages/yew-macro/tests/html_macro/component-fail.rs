use yew::html::ChildrenRenderer;
use yew::prelude::*;

#[derive(Clone, Properties, PartialEq)]
pub struct ChildProperties {
    #[prop_or_default]
    pub string: String,
    pub int: i32,
}

pub struct Child;
impl Component for Child {
    type Message = ();
    type Properties = ChildProperties;

    fn create(_ctx: &Context<Self>) -> Self {
        unimplemented!()
    }
    fn view(&self, _ctx: &Context<Self>) -> Html {
        unimplemented!()
    }
}

#[derive(Clone, Properties, PartialEq)]
pub struct ChildContainerProperties {
    pub children: ChildrenWithProps<Child>,
}

pub struct ChildContainer;
impl Component for ChildContainer {
    type Message = ();
    type Properties = ChildContainerProperties;

    fn create(_ctx: &Context<Self>) -> Self {
        unimplemented!()
    }
    fn view(&self, _ctx: &Context<Self>) -> Html {
        unimplemented!()
    }
}

fn compile_fail() {
    html! { <Child> };
    html! { <Child:: /> };
    html! { <Child with /> };
    html! { <Child .. /> };
    html! { <Child ..{ 5 + } /> };
    html! { <Child props /> };
    html! { <Child with props > };
    html! { <Child ..props > };
    let (p1, p2);
    html! { <Child with p1 with p2 /> };
    html! { <Child ..p1 ..p2 /> };
    html! { <Child with props ref={()} ref={()} /> };
    html! { <Child ..props ref={()} ref={()} /> };
    html! { <Child with props ref={()} ref={()} value=1 /> };
    html! { <Child ..props ref={()} ref={()} value=1 /> };
    html! { <Child with props ref={()} value=1 ref={()} /> };
    html! { <Child ..props ref={()} value=1 ref={()} /> };
    html! { <Child with props value=1 ref={()}  ref={()} /> };
    html! { <Child ..props value=1 ref={()}  ref={()} /> };
    html! { <Child value=1 with props  ref={()}  ref={()} /> };
    html! { <Child value=1 ..props  ref={()}  ref={()} /> };
    html! { <Child value=1 ref={()} with props ref={()} /> };
    html! { <Child value=1 ref={()} ..props ref={()} /> };
    html! { <Child ref={()} ref={()} value=1  with props  /> };
    html! { <Child ref={()} ref={()} value=1 ..props  /> };
    html! { <Child ..blah /> };
    html! { <Child value=1 ..props /> };
    html! { <Child .. props value=1 /> };
    html! { <Child type=0 /> };
    html! { <Child ref=() /> };
    html! { <Child invalid-prop-name=0 /> };
    html! { <Child unknown="unknown" /> };
    html! { <Child string= /> };
    html! { <Child int=1 int=2 int=3 /> };
    html! { <Child int=1 string={} /> };
    html! { <Child int=1 string=3 /> };
    html! { <Child int=1 string={3} /> };
    html! { <Child int=1 ref={()} /> };
    html! { <Child int=1 ref={()} ref={()} /> };
    html! { <Child int=0u32 /> };
    html! { <Child string="abc" /> };
    html! { </Child> };
    html! { <Child><Child></Child> };
    html! { <Child></Child><Child></Child> };
    html! { <Child>{ "Not allowed" }</Child> };

    let num = 1;
    html! { <Child int=num ..props /> };

    // trying to overwrite `children` on props which don't take any.
    html! {
        <Child ..ChildProperties { string: "hello".to_owned(), int: 5 }>
            { "please error" }
        </Child>
    };

    html! { <ChildContainer /> };
    html! { <ChildContainer></ChildContainer> };
    html! { <ChildContainer>{ "Not allowed" }</ChildContainer> };
    html! { <ChildContainer><></></ChildContainer> };
    html! { <ChildContainer><other /></ChildContainer> };

    // using `children` as a prop while simultaneously passing children using the syntactic sugar
    let children = ChildrenRenderer::new(vec![html_nested! { <Child int=0 /> }]);
    html! {
        <ChildContainer {children}>
            <Child int=1 />
        </ChildContainer>
    };

    html_nested! {
        <span>{ 1 }</span>
        <span>{ 2 }</span>
    };

    html! { <Child {std::f64::consts::PI} /> };
    html! { <Child {7 + 6} /> };
    html! { <Child {children.len()} /> };

}

fn main() {}
