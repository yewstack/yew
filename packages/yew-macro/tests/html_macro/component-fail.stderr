error: this opening tag has no corresponding closing tag
  --> $DIR/component-fail.rs:43:13
   |
43 |     html! { <Child> };
   |             ^^^^^^^

error: unexpected end of input, expected identifier
  --> $DIR/component-fail.rs:44:13
   |
44 |     html! { <Child:: /> };
   |             ^^^^^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:45:20
   |
45 |     html! { <Child with /> };
   |                    ^^^^

error: expected base props expression after `..`
  --> $DIR/component-fail.rs:46:20
   |
46 |     html! { <Child .. /> };
   |                    ^^

error: unexpected end of input, expected expression
  --> $DIR/component-fail.rs:46:13
   |
46 |     html! { <Child .. /> };
   |             ^^^^^^^^^^^^

error: expected base props expression after `..`
  --> $DIR/component-fail.rs:47:20
   |
47 |     html! { <Child ..{ 5 + } /> };
   |                    ^^

error: unexpected end of input, expected expression
  --> $DIR/component-fail.rs:47:22
   |
47 |     html! { <Child ..{ 5 + } /> };
   |                      ^^^^^^^

error: `props` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:48:20
   |
48 |     html! { <Child props /> };
   |                    ^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:49:20
   |
49 |     html! { <Child with props > };
   |                    ^^^^

error: this opening tag has no corresponding closing tag
  --> $DIR/component-fail.rs:50:13
   |
50 |     html! { <Child ..props > };
   |             ^^^^^^^^^^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:52:20
   |
52 |     html! { <Child with p1 with p2 /> };
   |                    ^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:54:20
   |
54 |     html! { <Child with props ref={()} ref={()} /> };
   |                    ^^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:55:20
   |
55 |     html! { <Child ..props ref={()} ref={()} /> };
   |                    ^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:56:20
   |
56 |     html! { <Child with props ref={()} ref={()} value=1 /> };
   |                    ^^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:57:20
   |
57 |     html! { <Child ..props ref={()} ref={()} value=1 /> };
   |                    ^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:58:20
   |
58 |     html! { <Child with props ref={()} value=1 ref={()} /> };
   |                    ^^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:59:20
   |
59 |     html! { <Child ..props ref={()} value=1 ref={()} /> };
   |                    ^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:60:20
   |
60 |     html! { <Child with props value=1 ref={()}  ref={()} /> };
   |                    ^^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:61:20
   |
61 |     html! { <Child ..props value=1 ref={()}  ref={()} /> };
   |                    ^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:62:28
   |
62 |     html! { <Child value=1 with props  ref={()}  ref={()} /> };
   |                            ^^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:63:28
   |
63 |     html! { <Child value=1 ..props  ref={()}  ref={()} /> };
   |                            ^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:64:37
   |
64 |     html! { <Child value=1 ref={()} with props ref={()} /> };
   |                                     ^^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:65:37
   |
65 |     html! { <Child value=1 ref={()} ..props ref={()} /> };
   |                                     ^^^^^^^

error: `with` doesn't have a value. (hint: set the value to `true` or `false` for boolean attributes)
  --> $DIR/component-fail.rs:66:47
   |
66 |     html! { <Child ref={()} ref={()} value=1  with props  /> };
   |                                               ^^^^

error: `ref` can only be specified once
  --> $DIR/component-fail.rs:67:20
   |
67 |     html! { <Child ref={()} ref={()} value=1 ..props  /> };
   |                    ^^^

error: base props expression must appear last in list of props
  --> $DIR/component-fail.rs:70:20
   |
70 |     html! { <Child .. props value=1 /> };
   |                    ^^^^^^^^

error: expected identifier, found keyword `type`
  --> $DIR/component-fail.rs:71:20
   |
71 |     html! { <Child type=0 /> };
   |                    ^^^^ expected identifier, found keyword
   |
help: you can escape reserved keywords to use them as identifiers
   |
71 |     html! { <Child r#type=0 /> };
   |                    ^^^^^^

error: the property value must be either a literal or enclosed in braces. Consider adding braces around your expression.
  --> $DIR/component-fail.rs:72:24
   |
72 |     html! { <Child ref=() /> };
   |                        ^^

error: expected a valid Rust identifier
  --> $DIR/component-fail.rs:73:20
   |
73 |     html! { <Child invalid-prop-name=0 /> };
   |                    ^^^^^^^^^^^^^^^^^

error: expected an expression following this equals sign
  --> $DIR/component-fail.rs:75:26
   |
75 |     html! { <Child string= /> };
   |                          ^

error: `int` can only be specified once but is given here again
  --> $DIR/component-fail.rs:76:26
   |
76 |     html! { <Child int=1 int=2 int=3 /> };
   |                          ^^^

error: `int` can only be specified once but is given here again
  --> $DIR/component-fail.rs:76:32
   |
76 |     html! { <Child int=1 int=2 int=3 /> };
   |                                ^^^

error: `ref` can only be specified once
  --> $DIR/component-fail.rs:81:26
   |
81 |     html! { <Child int=1 ref={()} ref={()} /> };
   |                          ^^^

error: this closing tag has no corresponding opening tag
  --> $DIR/component-fail.rs:84:13
   |
84 |     html! { </Child> };
   |             ^^^^^^^^

error: this opening tag has no corresponding closing tag
  --> $DIR/component-fail.rs:85:13
   |
85 |     html! { <Child><Child></Child> };
   |             ^^^^^^^

error: only one root html element is allowed (hint: you can wrap multiple html elements in a fragment `<></>`)
  --> $DIR/component-fail.rs:86:28
   |
86 |     html! { <Child></Child><Child></Child> };
   |                            ^^^^^^^^^^^^^^^

error: the property value must be either a literal or enclosed in braces. Consider adding braces around your expression.
  --> $DIR/component-fail.rs:90:24
   |
90 |     html! { <Child int=num ..props /> };
   |                        ^^^

error: cannot specify the `children` prop when the component already has children
   --> $DIR/component-fail.rs:108:26
    |
108 |         <ChildContainer {children}>
    |                          ^^^^^^^^

error: only one root html element is allowed (hint: you can wrap multiple html elements in a fragment `<></>`)
   --> $DIR/component-fail.rs:115:9
    |
115 |         <span>{ 2 }</span>
    |         ^^^^^^^^^^^^^^^^^^

error: only simple identifiers are allowed in the shorthand property syntax
   --> $DIR/component-fail.rs:118:21
    |
118 |     html! { <Child {std::f64::consts::PI} /> };
    |                     ^^^^^^^^^^^^^^^^^^^^

error: missing label for property value. If trying to use the shorthand property syntax, only identifiers may be used
   --> $DIR/component-fail.rs:119:21
    |
119 |     html! { <Child {7 + 6} /> };
    |                     ^^^^^

error: missing label for property value. If trying to use the shorthand property syntax, only identifiers may be used
   --> $DIR/component-fail.rs:120:21
    |
120 |     html! { <Child {children.len()} /> };
    |                     ^^^^^^^^^^^^^^

error[E0425]: cannot find value `blah` in this scope
  --> $DIR/component-fail.rs:68:22
   |
68 |     html! { <Child ..blah /> };
   |                      ^^^^ not found in this scope

error[E0425]: cannot find value `props` in this scope
  --> $DIR/component-fail.rs:69:30
   |
69 |     html! { <Child value=1 ..props /> };
   |                              ^^^^^ not found in this scope

error[E0308]: mismatched types
  --> $DIR/component-fail.rs:53:22
   |
53 |     html! { <Child ..p1 ..p2 /> };
   |              -----   ^^^^^^^ expected struct `ChildProperties`, found struct `std::ops::Range`
   |              |
   |              expected due to this
   |
   = note: expected struct `ChildProperties`
              found struct `std::ops::Range<_>`

error[E0308]: mismatched types
  --> $DIR/component-fail.rs:53:14
   |
53 |     html! { <Child ..p1 ..p2 /> };
   |              ^^^^^ expected struct `ChildProperties`, found struct `std::ops::Range`
   |
   = note: expected struct `ChildProperties`
              found struct `std::ops::Range<_>`

error[E0609]: no field `value` on type `ChildProperties`
  --> $DIR/component-fail.rs:69:20
   |
69 |     html! { <Child value=1 ..props /> };
   |                    ^^^^^ unknown field
   |
   = note: available fields are: `string`, `int`

error[E0609]: no field `r#type` on type `ChildProperties`
  --> $DIR/component-fail.rs:71:20
   |
71 |     html! { <Child type=0 /> };
   |                    ^^^^ unknown field
   |
   = note: available fields are: `string`, `int`

error[E0599]: no method named `r#type` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
  --> $DIR/component-fail.rs:71:20
   |
4  | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `r#type` not found for this
...
71 |     html! { <Child type=0 /> };
   |                    ^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`

error[E0609]: no field `unknown` on type `ChildProperties`
  --> $DIR/component-fail.rs:74:20
   |
74 |     html! { <Child unknown="unknown" /> };
   |                    ^^^^^^^ unknown field
   |
   = note: available fields are: `string`, `int`

error[E0599]: no method named `unknown` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
  --> $DIR/component-fail.rs:74:20
   |
4  | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `unknown` not found for this
...
74 |     html! { <Child unknown="unknown" /> };
   |                    ^^^^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`

error[E0277]: the trait bound `(): IntoPropValue<String>` is not satisfied
  --> $DIR/component-fail.rs:77:33
   |
77 |     html! { <Child int=1 string={} /> };
   |                                 ^^ the trait `IntoPropValue<String>` is not implemented for `()`

error[E0277]: the trait bound `{integer}: IntoPropValue<String>` is not satisfied
  --> $DIR/component-fail.rs:78:33
   |
78 |     html! { <Child int=1 string=3 /> };
   |                                 ^ the trait `IntoPropValue<String>` is not implemented for `{integer}`
   |
   = help: the following implementations were found:
             <&'static str as IntoPropValue<Cow<'static, str>>>
             <&'static str as IntoPropValue<Option<Cow<'static, str>>>>
             <&'static str as IntoPropValue<Option<String>>>
             <&'static str as IntoPropValue<String>>
           and 11 others

error[E0277]: the trait bound `{integer}: IntoPropValue<String>` is not satisfied
  --> $DIR/component-fail.rs:79:34
   |
79 |     html! { <Child int=1 string={3} /> };
   |                                  ^ the trait `IntoPropValue<String>` is not implemented for `{integer}`
   |
   = help: the following implementations were found:
             <&'static str as IntoPropValue<Cow<'static, str>>>
             <&'static str as IntoPropValue<Option<Cow<'static, str>>>>
             <&'static str as IntoPropValue<Option<String>>>
             <&'static str as IntoPropValue<String>>
           and 11 others

error[E0308]: mismatched types
  --> $DIR/component-fail.rs:80:31
   |
80 |     html! { <Child int=1 ref={()} /> };
   |                               ^^ expected struct `NodeRef`, found `()`

error[E0277]: the trait bound `u32: IntoPropValue<i32>` is not satisfied
  --> $DIR/component-fail.rs:82:24
   |
82 |     html! { <Child int=0u32 /> };
   |                        ^^^^ the trait `IntoPropValue<i32>` is not implemented for `u32`

error[E0599]: no method named `string` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
  --> $DIR/component-fail.rs:83:20
   |
4  | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `string` not found for this
...
83 |     html! { <Child string="abc" /> };
   |                    ^^^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`

error[E0609]: no field `children` on type `ChildProperties`
  --> $DIR/component-fail.rs:87:14
   |
87 |     html! { <Child>{ "Not allowed" }</Child> };
   |              ^^^^^ unknown field
   |
   = note: available fields are: `string`, `int`

error[E0599]: no method named `children` found for struct `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>` in the current scope
  --> $DIR/component-fail.rs:87:14
   |
4  | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `children` not found for this
...
87 |     html! { <Child>{ "Not allowed" }</Child> };
   |              ^^^^^ method not found in `ChildPropertiesBuilder<ChildPropertiesBuilderStep_missing_required_prop_int>`

error[E0609]: no field `children` on type `ChildProperties`
  --> $DIR/component-fail.rs:94:10
   |
94 |         <Child ..ChildProperties { string: "hello".to_owned(), int: 5 }>
   |          ^^^^^ unknown field
   |
   = note: available fields are: `string`, `int`

error[E0599]: no method named `build` found for struct `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>` in the current scope
  --> $DIR/component-fail.rs:99:14
   |
24 | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `build` not found for this
...
99 |     html! { <ChildContainer /> };
   |              ^^^^^^^^^^^^^^ method not found in `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>`

error[E0599]: no method named `build` found for struct `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>` in the current scope
   --> $DIR/component-fail.rs:100:14
    |
24  | #[derive(Clone, Properties, PartialEq)]
    |                 ---------- method `build` not found for this
...
100 |     html! { <ChildContainer></ChildContainer> };
    |              ^^^^^^^^^^^^^^ method not found in `ChildContainerPropertiesBuilder<ChildContainerPropertiesBuilderStep_missing_required_prop_children>`

error[E0277]: the trait bound `VChild<Child>: From<yew::virtual_dom::VText>` is not satisfied
   --> $DIR/component-fail.rs:101:31
    |
101 |     html! { <ChildContainer>{ "Not allowed" }</ChildContainer> };
    |                               ^^^^^^^^^^^^^ the trait `From<yew::virtual_dom::VText>` is not implemented for `VChild<Child>`
    |
    = note: required because of the requirements on the impl of `Into<VChild<Child>>` for `yew::virtual_dom::VText`
    = note: required by `into`

error[E0277]: the trait bound `VChild<Child>: From<VNode>` is not satisfied
   --> $DIR/component-fail.rs:102:29
    |
102 |     html! { <ChildContainer><></></ChildContainer> };
    |                             ^ the trait `From<VNode>` is not implemented for `VChild<Child>`
    |
    = note: required because of the requirements on the impl of `Into<VChild<Child>>` for `VNode`
    = note: required by `into`

error[E0277]: the trait bound `VChild<Child>: From<VNode>` is not satisfied
   --> $DIR/component-fail.rs:103:30
    |
103 |     html! { <ChildContainer><other /></ChildContainer> };
    |                              ^^^^^ the trait `From<VNode>` is not implemented for `VChild<Child>`
    |
    = note: required because of the requirements on the impl of `Into<VChild<Child>>` for `VNode`
    = note: required by `into`
